### PR TITLE
feat: add instrumentation and profiling endpoints

### PR DESCRIPTION
**Pull request by caitong93**
**What this PR does**

Add instrumentation and profiling endpoints

**why we need it**

**Special notes for your reviewer**:

### CODE REVIEW

File: examples/api-basic/main.go [hunk 1]
```diff
@@ -17,20 +17,35 @@ limitations under the License.
...
 
 func main() {
+	flag.Parse()
+
 	if err := web.RegisterDefaultEnvironment(); err != nil {
 		panic(err)
 	}
 
-	s := web.NewDefaultServer()
+	s, err := nirvana.New(&nirvana.Config{
```
Comment (Jimexist) [inline examples/api-basic/main.go]: maybe should be `server.New`?

File: metrics.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"github.com/caicloud/nirvana/web"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Metrics installs the default prometheus metrics handler
+type Metrics struct{}
+
+func (d Metrics) Install(s web.Server) {
+	if err := s.AddDescriptors(convertHandlerToDescriptor("/metrics", promhttp.Handler())); err != nil {
```
Comment (Jimexist) [inline metrics.go]: make the endpoint adjustable

**Comment by kdada** [comment]
Don't modify many things in a commit.

File: examples/api-basic/main.go (hunk 1 above)
Comment (caitong93) [reply examples/api-basic/main.go]: Do you mean a new package `github.com/caicloud/nirvana/server` ?

**Comment by caitong93** [comment]
@Jimexist updated

File: examples/api-basic/main.go [hunk 3]
```diff
@@ -17,20 +17,35 @@ limitations under the License.
...
+	"github.com/caicloud/nirvana"
+
 	"github.com/caicloud/nirvana/examples/api-basic/api/v1"
 	"github.com/caicloud/nirvana/examples/api-basic/api/v2"
+	"github.com/caicloud/nirvana/log"
 	"github.com/caicloud/nirvana/web"
 )
 
 func main() {
+	flag.Parse()
```
Comment (ddysher) [inline examples/api-basic/main.go]: why do you need `flag.Parse`?

BTW, if you do need flag, then use cli package :)

File: examples/api-basic/main.go [hunk 4]
```diff
@@ -17,20 +17,35 @@ limitations under the License.
 package main
 
 import (
+	"flag"
```
Comment (ddysher) [inline examples/api-basic/main.go]: I'd prefer using a separate example for metrics and profiling, to make it more focused.

File: metrics.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+// Metrics installs the default prometheus metrics handler
+type Metrics struct{}
+
+func (d Metrics) Install(s web.Server, metricsPath string) {
+	if metricsPath == "" {
+		metricsPath = "/metrics"
+	}
+	if err := s.AddDescriptors(convertHandlerToDescriptor(metricsPath, promhttp.Handler())); err != nil {
```
Comment (ddysher) [inline metrics.go]: `convertHandlerToDescriptor` 👈  looks like we can create some syntax sugar for this?  @caitong93 @kdada

File: nirvana.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		s.SetLogger(c.Logger)
+	}
+	installAPI(s, c)
+	return s, nil
+}
+
+// installAPI installs additional APIs used for debuging, instrumentation ...
+func installAPI(s web.Server, c *Config) {
+	if c.EnableProfiling {
+		Profiling{}.Install(s)
```
Comment (ddysher) [inline nirvana.go]: Do I have to instantiate a `Profiling` struct?   It looks like we can just use exported method?

File: nirvana.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		if c.EnableContentionProfiling {
+			runtime.SetBlockProfileRate(1)
+		}
+	}
+	if c.EnableMetrics {
+		Metrics{}.Install(s, c.MetricsPath)
```
Comment (ddysher) [inline nirvana.go]: ditto

File: metrics.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
+/*
```
Comment (ddysher) [inline metrics.go]: Do we want to put `metrics.go` and `profiling.go` at top level?

File: nirvana.go (hunk 6 above)
Comment (caitong93) [reply nirvana.go]: > Do I have to instantiate a Profiling struct?
No





> It looks like we can just use exported method?

Yes

File: metrics.go (hunk 5 above)
Comment (caitong93) [reply metrics.go]: I have discussed that with @kdada . Sugar is important, so we better have another pr for that.

File: examples/api-basic/main.go (hunk 3 above)
Comment (caitong93) [reply examples/api-basic/main.go]: This is a mistake. Should be removed.

File: log/glog.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
+/*
```
[the remark this answers is not in the document]
Comment (caitong93) [reply log/glog.go]: OK

File: metrics.go (hunk 8 above)
Comment (caitong93) [reply metrics.go]: If anyone else have an idea where is a suitable place for these files, then we can move them.

**Comment by caitong93** [comment]
@ddysher @kdada @Jimexist 
updated, PTAL
Can I rebase now

**Comment by ddysher** [comment]
We need an offline discussion around instrumentation.

**Comment by Jimexist** [comment]
@caitong93 please rebase/merge from `master` first

**Comment by ddysher** [comment]
Can you make it next Wendesday?

**Comment by caitong93** [comment]
@ddysher @kdada @Jimexist rebased

**Comment by ddysher** [comment]
@kdada to review

**Pull request merged**